R/module-helpers.R

```diff
@@ -353,3 +353,120 @@ get_ssp_regions <- function(pcGDP, reg_names, income_group,
+#'   - Extrapolated using an exponential decay function paramaterized by the columns
+#'     \code{improvement.rate} and \code{improvement.max} using the following formulation
+#'     \code{v_0*max+(v_0-v_0*max)*(1-rate)^(y-y_0)}
+#'   - For rows that specify a char value in the column \code{improvement.shadow.technology}
+#'     exponential decay will be calculated on the difference between the values calculated
+#'     by left joining with itself on the column \code{improvement.shadow.technology} with
+#'     the column \code{technology}.  In other words for shadowing technologies the decay is
+#'     only applied to the difference in the values in the last year in which one was
+#'     specified. This is to allow for instance a Gas CC plant to have cost reductions at a
+#'     moderate pace but a Gas CC+CCS can have rapid cost reductions tothe CCS portion of
+#'     the cost.
+#'
+#' @param d The wide format tibble with values under year columns that will be filled
+#' to ensure the given years are present using the rules mentioned above.
+#' @param out_years A vector of integers specifying which years must have values in the
+#' output data.
+#' @return The filled out data set in long format.  The years will be in the \code{year}
+#' column and will include all values in \code{out_years} and the filled in values will
+#' be in the \code{value} column.  All extrapolation parameters will be cleaned out.
+#' @importFrom tibble has_name
+#' @importFrom dplyr filter mutate select group_by_ setdiff rename ungroup
+#' @importFrom tidyr gather complete nesting_
+#' @importFrom assertthat assert_that
+#' @author Pralit Patel
+fill_exp_decay_extrapolate <- function(d, out_years) {
+  . <- value <- year <- improvement.rate <- improvement.max <-
+    improvement.shadow.technology <- technology <- year_base <-
+    value_base <- shadow.value <- NULL  # silence package check notes
+
+  # Some error checking first
+  assert_that(has_name(d, "improvement.rate"))
+  assert_that(has_name(d, "improvement.max"))
+  # either improvement.rate/max are both NA or neither are
+  assert_that(all(is.na(d$improvement.rate) == is.na(d$improvement.max)))
+
+  # Note we want to allow use of the improvement.shadow.technology feature to be
+  # optional so we will check if they have not provided the column and fill NAs
+  # if not to avoid error
+  d %>%
+    mutate(improvement.shadow.technology =
+             if(has_name(., "improvement.shadow.technology")) improvement.shadow.technology else as.character(NA)) ->
+    d
+
+  # The first step is to linearly interpolate missing values that are in between
+  # values which are specified (approx_fun rule=1)
+  d %>%
+    gather(year, value, matches(YEAR_PATTERN)) %>%
+    mutate(year = as.integer(year)) %>%
+    complete(nesting_(select(., -year, -value)), year = union(year, out_years)) %>%
+    group_by_(.dots=paste0('`',names(.)[!(names(.) %in% c("year", "value"))], '`')) %>%
```

Comment:
Done.  But do you have any suggestions for doing the `group_by` better?  I feel like this comes up often: I would like to group by everything except two columns such as a `year` and `value` but there doesn't appear to be any good shorthand for doing this (it seems maybe they have added something in dplyr 0.7 but I'm not sure if folks are even using that, I certainly am not).